src/fastfield/serializer/mod.rs

```diff
@@ -0,0 +1,164 @@
+use crate::common::BinarySerializable;
+use crate::common::CompositeWrite;
+use crate::common::CountingWriter;
+use crate::directory::WritePtr;
+use crate::schema::Field;
+use fastfield_codecs::CodecId;
+//pub use bitpacked::BitpackedFastFieldSerializer;
+pub use fastfield_codecs::bitpacked::BitpackedFastFieldSerializer;
+use fastfield_codecs::linearinterpol::LinearInterpolFastFieldSerializer;
+pub use fastfield_codecs::FastFieldDataAccess;
+pub use fastfield_codecs::FastFieldSerializerEstimate;
+pub use fastfield_codecs::FastFieldStats;
+use std::io::{self, Write};
+
+/// `CompositeFastFieldSerializer` is in charge of serializing
+/// fastfields on disk.
+///
+/// Fast fields have different encodings like bit-packing.
+///
+/// `FastFieldWriter`s are in charge of pushing the data to
+/// the serializer.
+/// The serializer expects to receive the following calls.
+///
+/// * `new_u64_fast_field(...)`
+/// * `add_val(...)`
+/// * `add_val(...)`
+/// * `add_val(...)`
+/// * ...
+/// * `close_field()`
+/// * `new_u64_fast_field(...)`
+/// * `add_val(...)`
+/// * ...
+/// * `close_field()`
+/// * `close()`
+pub struct CompositeFastFieldSerializer {
+    composite_write: CompositeWrite<WritePtr>,
+}
+
+impl CompositeFastFieldSerializer {
+    /// Constructor
+    pub fn from_write(write: WritePtr) -> io::Result<CompositeFastFieldSerializer> {
+        // just making room for the pointer to header.
+        let composite_write = CompositeWrite::wrap(write);
+        Ok(CompositeFastFieldSerializer { composite_write })
+    }
+
+    /// Serialize data into a new u64 fast field. The compression will be detected automatically.
+    pub fn create_auto_detect_u64_fast_field(
+        &mut self,
+        field: Field,
+        stats: FastFieldStats,
+        fastfield_accessor: impl FastFieldDataAccess,
+        data_iter_1: impl Iterator<Item = u64>,
+        data_iter_2: impl Iterator<Item = u64>,
+    ) -> io::Result<()> {
+        let field_write = self.composite_write.for_field_with_idx(field, 0);
+
+        let mut estimations = vec![];
+
+        {
+            let (ratio, name, id) = (
+                BitpackedFastFieldSerializer::<Vec<u8>>::estimate(
+                    &fastfield_accessor,
+                    stats.clone(),
+                ),
+                BitpackedFastFieldSerializer::<Vec<u8>>::NAME,
+                BitpackedFastFieldSerializer::<Vec<u8>>::ID,
+            );
+            estimations.push((ratio, name, id));
+        }
+        {
+            let (ratio, name, id) = (
+                LinearInterpolFastFieldSerializer::estimate(&fastfield_accessor, stats.clone()),
+                LinearInterpolFastFieldSerializer::NAME,
+                LinearInterpolFastFieldSerializer::ID,
+            );
+            estimations.push((ratio, name, id));
+        }
+        estimations.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
+        let (_ratio, name, id) = estimations[0];
```

Comment:
max_by also exists.

Reply:
we need sort_by eventually anyway when the estimate needs to be big enough in difference to the other ones to justify a slower algorithm

Reply:
won't it be a max_by a linear some function between  cost and benefit?

Reply:
hmm yes probably